### PR TITLE
feat 1588 Add additional validation rules to new price flow

### PR DESCRIPTION
**Pull request by ShaBloze**
## Description

A few charge price validation rules is missing and needs to be added to the new price flow.

Following rules are missing:

- ResolutionValidationRule 
- StartDateValidationRule 
- PreviousOperationsMustBeValidRule 
- ResolutionIsRequired 

## References

* #1588

### CODE REVIEW

File: docs/business-processes/asynchronous-validations.md [hunk 1]
```diff
@@ -20,9 +20,9 @@ The following asynchronous validation rules are currently implemented in the cha
 |VR.449|The type of a charge has domain values D01 (Subscription), D02 (Fee), D03 (Tariff)|E86|✓|✓||
 |VR.457|The energy price of a charge consists of maximal 14 digits with format 99999999.999999|E86|✓|✓||
 |VR.488|The VAT class of a charge has domain values D01 (No VAT), D02 (VAT)|E86|✓|||
-|VR.505-1|The Tariff to which the charge information applies must have period type Day, Hour or Quarter of Hour|D23|(✓) Tariff only|||
-|VR.505-2|The Fee to which the charge information applies must have period type Month|D23|(✓) Fee only|||
-|VR.505-3|The Subscription to which the charge information applies must have period type Month|D23|(✓) Subscription only|||
+|VR.505-1|The Tariff to which the charge information applies must have period type Day, Hour or Quarter of Hour|D23|(✓) Tariff only|✓||
+|VR.505-2|The Fee to which the charge information applies must have period type Month|D23|(✓) Fee only|✓||
+|VR.505-3|The Subscription to which the charge information applies must have period type Month|D23|(✓) Subscription only|✓||
```
Comment (prtandrup) [inline docs/business-processes/asynchronous-validations.md:25]: Suggest below text naming so it works for both information and prices

```suggestion
|VR.505-1|Tariff must have period type Day, Hour or Quarter of Hour|D23|(✓) Tariff only|(✓) Tariff only||
|VR.505-2|Fee must have period type Month|D23|(✓) Fee only|(✓) Fee only||
|VR.505-3|Subscription must have period type Month|D23|(✓) Subscription only|(✓) Subscription only||
```

File: docs/business-processes/asynchronous-validations.md [hunk 2]
```diff
@@ -9,7 +9,7 @@ The following asynchronous validation rules are currently implemented in the cha
 |VR.153|The recipient of a message is mandatory|D02|✓|✓||
 |VR.165|The recipient role of a message must be DDZ (Metering point administrator or DataHub)|E55|✓|✓||
 |VR.200|Metering point do not exist|E10|||✓|
-|VR.209|The data information must be received within the correct time period|E17|✓|||
+|VR.209|The data information must be received within the correct time period|E17|✓|✓||
```
Comment (prtandrup) [inline docs/business-processes/asynchronous-validations.md:12]:
```suggestion
|VR.209|The data received must be within the correct time period|E17|✓|✓||
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/BusinessValidation/ValidationRules/PreviousOperationsMustBeValidRule.cs [hunk 3]
```diff
@@ -18,9 +18,9 @@ namespace GreenEnergyHub.Charges.Domain.Dtos.ChargeInformationCommands.Validatio
 {
     public class PreviousOperationsMustBeValidRule : IValidationRuleWithExtendedData
     {
-        public PreviousOperationsMustBeValidRule(string triggeredBy)
+        public PreviousOperationsMustBeValidRule(ChargeInformationOperationDto chargeInformationOperationDto)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/BusinessValidation/ValidationRules/PreviousOperationsMustBeValidRule.cs:21]: If we do this, could we call the parameter `rejectedOperation` or `previousOperation` something, so it's easier to figure out what goes where, the next time we visit this.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/BusinessValidation/ValidationRules/PreviousOperationsMustBeValidRule.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using GreenEnergyHub.Charges.Domain.Dtos.Validation;
+
+namespace GreenEnergyHub.Charges.Domain.Dtos.ChargePriceCommands.Validation.BusinessValidation.ValidationRules
+{
+    public class PreviousOperationsMustBeValidRule : IValidationRuleWithExtendedData
+    {
+        public PreviousOperationsMustBeValidRule(ChargePriceOperationDto chargePriceOperationDto)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/BusinessValidation/ValidationRules/PreviousOperationsMustBeValidRule.cs:21]: Similar to earlier comment; rejectedOperation or previousOperation?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/ResolutionIsRequiredRule.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+using GreenEnergyHub.Charges.Domain.Charges;
+using GreenEnergyHub.Charges.Domain.Dtos.Validation;
+
+namespace GreenEnergyHub.Charges.Domain.Dtos.ChargePriceCommands.Validation.InputValidation.ValidationRules
+{
+    public class ResolutionIsRequiredRule : IValidationRule
+    {
+        private readonly ChargePriceOperationDto _chargeInformationOperationDto;
+
+        public ResolutionIsRequiredRule(ChargePriceOperationDto chargeInformationOperationDto)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/ResolutionIsRequiredRule.cs:24]:
```suggestion
        private readonly ChargePriceOperationDto _chargePriceOperationDto;

        public ResolutionIsRequiredRule(ChargePriceOperationDto chargePriceOperationDto)
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/ResolutionFeeValidationRule.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+namespace GreenEnergyHub.Charges.Domain.Dtos.ChargePriceCommands.Validation.InputValidation.ValidationRules
+{
+    public class ResolutionFeeValidationRule : ResolutionValidationRule
+    {
+        public ResolutionFeeValidationRule(ChargePriceOperationDto chargePriceOperationDto)
+            : base(
+                chargePriceOperationDto,
+                ChargeType.Fee,
+                new List<Resolution> { Resolution.P1M },
+                Dtos.Validation.ValidationRuleIdentifier.ResolutionFeeValidation)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/ResolutionFeeValidationRule.cs:27]: Could we add below using and remove this Dtos.Validation

`using GreenEnergyHub.Charges.Domain.Dtos.Validation;`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/ResolutionTariffValidationRule.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+namespace GreenEnergyHub.Charges.Domain.Dtos.ChargePriceCommands.Validation.InputValidation.ValidationRules
+{
+    public class ResolutionTariffValidationRule : ResolutionValidationRule
+    {
+        public ResolutionTariffValidationRule(ChargePriceOperationDto chargePriceOperationDto)
+            : base(
+                chargePriceOperationDto,
+                ChargeType.Tariff,
+                new List<Resolution> { Resolution.P1D, Resolution.PT1H, Resolution.PT15M },
+                Dtos.Validation.ValidationRuleIdentifier.ResolutionTariffValidation)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/ResolutionTariffValidationRule.cs:27]: Could we add below using and remove this Dtos.Validation

`using GreenEnergyHub.Charges.Domain.Dtos.Validation;`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/StartDateValidationRule.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        // -720 lower limit is only for test environments. In production this should be +31 days
+        private const int LowerTimeLimitInDays = -720;
+        private const int UpperTimeLimitInDays = 1095;
+        private readonly Instant _validityStartDate;
+        private readonly Instant _periodStart;
+        private readonly Instant _periodEnd;
+
+        public StartDateValidationRule(
+            ChargePriceOperationDto chargePriceOperationDto,
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/StartDateValidationRule.cs:31]: I like that it takes the Dto as param, could you perhaps also change the StartDateValidationRule for Information so it take the informationDto as well?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ResolutionFeeValidationRuleTests.cs [hunk 9]
```diff
@@ -103,8 +103,13 @@ public class ResolutionFeeValidationRuleTests
         [InlineAutoDomainData]
         public void ValidationRuleIdentifier_ShouldBe_EqualTo(ChargeInformationOperationDtoBuilder chargeInformationOperationDtoBuilder)
         {
+            // Arrange
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ValidationRules/ResolutionFeeValidationRuleTests.cs:106]: Fyi, when test spans 3 lines only, we are not obligated to use AAA

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/StartDateValidationRuleTests.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace GreenEnergyHub.Charges.Tests.Domain.Dtos.ChargePriceCommands.Validation.InputValidation.ValidationRules
+{
+    [UnitTest]
+    public class StartDateValidationRuleTests
+    {
+        [Theory]
+        [InlineAutoMoqData(-1000, false)]
+        [InlineAutoMoqData(0, true)]
+        [InlineAutoMoqData(2000, false)]
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/StartDateValidationRuleTests.cs:38]: Think we can improve the test ( and the version for imformation ) by doing boundary value testing like this:

```suggestion
        [InlineAutoMoqData(-721, false)]
        [InlineAutoMoqData(720, true)]
        [InlineAutoMoqData(0, true)]
        [InlineAutoMoqData(1095, true)]
        [InlineAutoMoqData(1096, false)]
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/BusinessValidation/ValidationRules/PreviousOperationsMustBeValidRule.cs (hunk 3 above)
Comment (ShaBloze) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeInformationCommands/Validation/BusinessValidation/ValidationRules/PreviousOperationsMustBeValidRule.cs:21]: Changed it to `previousOperation`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/BusinessValidation/ValidationRules/PreviousOperationsMustBeValidRule.cs (hunk 4 above)
Comment (ShaBloze) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/BusinessValidation/ValidationRules/PreviousOperationsMustBeValidRule.cs:21]: Changed it to `previousOperation`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/StartDateValidationRule.cs (hunk 8 above)
Comment (ShaBloze) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargePriceCommands/Validation/InputValidation/ValidationRules/StartDateValidationRule.cs:31]: Yes of course! Changed it for the StartDateValidationRule for Information as well.

**Review by prtandrup — APPROVED** [review]
LGTM 👍
